### ISSUE TITLE
Allow combining multiple animations

### ISSUE THREAD

**Issue by appgurueu** [minetest/minetest#12663]
## Problem

The model file formats allow setting specific animations. A model can however only have exactly one animation at a time. This makes it impossible to combine many animations decently. One example of this is the player model used by MTG: Walking & mining are independent. The straightforward workaround here was to bake animations for all 4 possible combinations into the model:

* Neither walking nor mining: `stand`
* Walking but not mining: `walk`
* Not walking but mining: `mine`
* Walking and mining: `walk_mine`

For a few animations that can be combined this obviously works well, but for multiple independent animations a combinatorial explosion would soon cause things to blow up in terms of media size (due to number of animations) as well as making it harder for model creators to create the models (AFAICT there is no tool to generate all combinations of a set of animations for you; it's currently done manually).

## Solutions

Allow combining multiple animations: Replace the single animation with an animation stack:

* `object:push_animation(frame_range, frame_speed, frame_blend, frame_loop)`
* `frame_range, frame_speed, frame_blend, frame_loop = object:pop_animation()`

For backwards compatibility, `frame_range, frame_speed, frame_blend, frame_loop = object:get_animation()` would return the topmost animation on the stack, and `object:set_animation(frame_range, frame_speed, frame_blend, frame_loop)` would replace the topmost animation.

The animations would then be combined *in that order*, from top to bottom, to make the final animation. In most cases the order of application probably won't matter anyways since the animations will affect entirely different bones.

Implementation of this will require changes to both IrrlichtMT and Minetest.

## Alternatives

One alternative is to fully animate models serverside through overriding *all* bones as character_anim does by using modlib's b3d reader. The obvious downside of this is that it isn't nearly as smooth.

Another far-fetched alternative would be fully animating models *clientside*. This requires extensions to the CSM API and ideally SSCSM.

## Additional context

Related: Frame blending (unfortunately broken).

**Comment by KaylebJay** [comment]
I could use this feature too - if you have an animal that you want to have head animations with, and you also have walking, running animations, at the moment you have to make a new animation for every single head movement depending on the animals state (walking, running, standing)!

**Comment by appgurueu** [comment]
> I could use this feature too - if you have an animal that you want to have head animations with, and you also have walking, running animations, at the moment you have to make a new animation for every single head movement depending on the animals state (walking, running, standing)!

Yeah, or you use bones, which gives you more flexibility, but is also much jankier (usually too janky).

I wish we had (SS)CSM capable of controlling bones.

**Comment by Desour** [comment]
Another API that would probably work equally well (at least if bone parents are preserved), would be to set animation for only a set of bones.

**Comment by appgurueu** [comment]
I think if we don't want to require passing a set of bones, this will depend on the file format supporting "multiple" animations cleanly (glTF does):
Each animation defines a bunch of channels overriding bone properties. If you don't want to animate a bone, simply don't have a channel targeting it. And even if your model editor were not to cooperate, you can relatively easily edit this in the JSON.

For frame ranges, it's nastier: Should we pretend that a frame range doesn't animate a bone when it has no frames in the given range? When the "overrides" in the given range (exactly?) equal the static bone properties? This seems hacky and error prone.

In the end I think we can and probably should have both the ability to specify a set of bones and to infer this from the glTF animation. New proposal:

```
-- note: new API method. could alternatively overload set_animation but that ends up dirtier.
obj:set_anim({
  {
    name = "anim name", -- alternatively: 1-based index, defaulting to 1
    from = 0, -- start time
    to = math.huge, -- clamped to actual anim length on the client
    speed = 1, -- either a default of 1 or make it mandatory
    loop = false,
    bones = {name = true, ...}
  },
  ...
})
```
